Update validate-autoinstall-user-data script

Chris-Peterson444 opened the pull request:
./scripts/validate-autoinstall-user-data is used by the integration tests to verify the written user data is correct, but we can additionally advertise this script as something for users to pre-validate their autoinstall data. 

Changes include some refactoring and trying to catch some likely common mistakes.

Todo:
- [x] Write how-to page in documentation
- ~~[ ] Tests~~

dbungert commented on scripts/validate-autoinstall-user-data.py:
```diff
@@ -29,56 +29,117 @@
+        "--check-link",
+        dest="check_link",
+        action="store_true",
+        help="Assert the documentation link is in the user data",
```
I suggest help=argparse.SUPPRESS, as this is not something I expect users to be concerned about and may instead just be confusing.  I wouldn't want people to think that this is a requirement of their autoinstall data.  Please replace with a comment explaining that we just want the link in autoinstall-user-data.

Chris-Peterson444 replied:
Good point, done. Thanks!

Chris-Peterson444 commented:
I think this is ready. It's not perfect but it's a good start. I think I would like to punt writing tests for this since it uses the dry-run logic that is already tested. Eventually I think we could use this in CI for some more testing, which at that point we could write some more tests for this particular script if we wanted. 

Additionally, I would like to package this (and subiquity as a whole) in a way that users could just install the subiquity snap and invoke the validator from there, which would reduce the setup here.

ogayot commented on scripts/validate-autoinstall-user-data.py:
```diff
@@ -27,64 +27,224 @@
+    with tempfile.TemporaryDirectory() as td:
+        path = Path(td) / "autoinstall.yaml"
+        with open(path, "w") as tf:
+            tf.write(ai_data)
```
suggestion: 

```python
path.write_text(ai_data)
```

https://docs.python.org/3/library/pathlib.html#pathlib.Path.write_text

Chris-Peterson444 replied:
TIL. Thanks for pointing that out!

ogayot commented on scripts/validate-autoinstall-user-data.py:
```diff
@@ -27,64 +27,224 @@
+        default=False,
+    )
+
+    args: Namespace = parser.parse_args()
```
return parser.parse_args() ?

ogayot commented on scripts/validate-autoinstall-user-data.py:
```diff
@@ -27,64 +27,224 @@
+            tf.write(ai_data)
 
-    data = yaml.safe_load(user_data)
+        ret_code = asyncio.run(
```
return asyncio.run(...) ?

ogayot commented on scripts/validate-autoinstall-user-data.py:
```diff
@@ -27,64 +27,224 @@
+        except Exception as exc:
+            print(f"{type(exc).__name__}: {exc}")
+            print(FAILURE_MSG)
+            sys.exit(1)
```
return 1?

ogayot commented on doc/howto/autoinstall-validation.rst:
```diff
@@ -0,0 +1,249 @@
+
+.. code:: none
+
+   git clone git@github.com:canonical/subiquity.git && cd subiquity/
```
I'd recommend cloning with the https protocol, rather than SSH ; so it works for everyone.

ogayot commented on doc/howto/autoinstall-validation.rst:
```diff
@@ -0,0 +1,249 @@
+   finish: subiquity/load_autoinstall_config: FAIL: Malformed autoinstall in 'version or interactive-sections' section
+   Malformed autoinstall in 'version or interactive-sections' section
+   Traceback (most recent call last):
+     File "/home/cpete/software/subiquity/scripts/../subiquity/server/server.py", line 654, in validate_autoinstall
```
I'd typically drop the "/home/cpete/software/subiquity" prefix from the tracebacks but it's up to you

Chris-Peterson444 replied:
Good catch, I hadn't thought about that. I replaced the real prefix with ellipses. Does that seem better?

ogayot commented on scripts/validate-autoinstall-user-data.py:
```diff
@@ -27,64 +27,224 @@
-    # Verify autoinstall doc link is in the file
+        print(exc)  # Has the useful error message
+
+        # Print the full traceback if verbosity >=2
```
> 2

ogayot commented on scripts/validate-autoinstall-user-data.py:
```diff
@@ -27,64 +27,224 @@
+
+
+SUCCESS_MSG = "Success: The provided autoinstall config validated succesfully"
+FAILURE_MSG = "Failure: The provided autoinstall config did not validate succesfully"
```
nitpick: I'd reword just a bit so that the last words written are not "validate successfully". I'm a lazy reader :laughing: 

The provided autoinstall config failed validation?

ogayot replied:
Also, I'm tempted to suggest `rich` (https://rich.readthedocs.io/en/stable/introduction.html) but it's probably difficult to justify for only one message.

Chris-Peterson444 replied:
> lazy reader

Me too. I couldn't come up with something at the time but I like it. I used your suggested text. 

> I'm tempted to suggest `rich`

I do like the idea. And it's in Main, which is good. Although maybe once (if) we start having more error messaging we could use it?

ogayot replied:
yeah, let's think about it later. This is not blocking.

ogayot commented on scripts/validate-autoinstall-user-data.py:
```diff
@@ -27,64 +27,224 @@
+# At the very least, local curtin needs to be in the front of the python path
+sys.path.insert(0, str(subiquity_root))
+sys.path.insert(1, str(curtin_root))
+sys.path.insert(2, str(probert_root))
```
I don't love this but I don't have any better suggestions, so :laughing:

Chris-Peterson444 replied:
I agree, it's not great. I can't yet think of any less worse alternatives, except for packaging it into the snap somehow.

ogayot reviewed (commented):
Looks nice overall. A few comments though.

Chris-Peterson444 commented:
Updated + rebased onto main.

Thanks for the feedback @ogayot

ogayot reviewed (commented):
LGTM once we can make the CI happy.

Chris-Peterson444 commented:
Finally coming back to this. A couple of things:

1. I've refactored the code a bit and split the changes into smaller logical
commits. Hopefully this makes re-review a little easier if necessary.

2. Changing the CI to rely on the new behavior isn't going to work right away.
It's uncovering quite a few things we should eventually fix, but I don't think
we need to do that all in this PR. I'm proposing a hidden `--legacy` flag we
can use in CI to rely on the old JSON schema direct validation behavior and
over time we can work out the fixes to make CI happy with the dry-run validation.
What do you think?

dbungert reviewed (approved):
An interesting challenge with the run-time verifier is that it has assumptions that may not match the target installation ISO.
Examples, which can be valid but will fail validation:
* missing identity / user-data - would be valid on Desktop
* source asking for anything but `id: synthesized` - might be valid, if the id matches install-sources.yaml on the ISO

I feel like the validator is still useful but I think it would be worth mentioning these limitations, if we haven't already.

Chris-Peterson444 commented:
Great point. This is one of the issues I was trying to avoid in CI for now and we should definitely document it. How does this look? I added a new section "Validator Limitations" and tried to address all of the things I might consider as limitations of the validator (mostly rearranging the disclaimers I had put in the intro previously), including the points you provided.
